src/app/notes/state/notes.reducer.ts

```diff
@@ -1,22 +1,30 @@
 import { createReducer, on } from '@ngrx/store';
 import { NotesState } from '../interface/note';
-import { addNote } from './notes.actions';
-import { updateNote} from './notes.actions';
-
-
+import { addOrUpdateNote } from './notes.actions';
 
 export const initialState: NotesState = {
   notes: [],
 };
 
 export const notesReducer = createReducer(
   initialState,
-  on(addNote, (state, { note }) => ({
-    ...state,
-    notes: [...state.notes, note],
-  })),
-  on(updateNote, (state, { note }) => ({
-    ...state,
-    notes: state.notes.map((n) => (n.title.toLowerCase() === note.title.toLocaleLowerCase() ? note : n)),
-  }))
-);
\ No newline at end of file
+  on(addOrUpdateNote, (state, { note }) => {
+    const notes = [...state.notes];
+    const existingNoteIndex = state.notes.findIndex(
+      (n) => n.title.toLowerCase() === note.title.toLowerCase()
+    );
+
+    if (existingNoteIndex > -1) {
+      notes[existingNoteIndex] = note;
+      alert('Note has been updated successfully!');
```

Comment:
We could move the alerting to an effect if we wanted to separate concerns further, but for this simple example I kept it with the update logic.
Another alternative is storing the alert message in the state, and alerting when the state changes.